### DISCUSSION TITLE
Lists of minimum or maximum size

### DISCUSSION THREAD

**Discussion by tnarg** [cuelang/cue#401] [General] [answered]
In the [List tour](https://cuelang.org/docs/tutorials/tour/types/lists/) I see that you can specify lists of exact length, e.g. `4 * [uint8]`. Is it possible to specify a minimum or maximum length for lists? I could specify a uint8 List with between 1 and 4 elements today by enumerating the possibilities like so

```
1 * [uint8] | 2 * [uint8] | 3 * [uint8] | 4 * [uint8]
```

For such a small range, this isn't too bad, but for a list between 0 and 50 elements, this becomes exceedingly verbose.

**Comment by tnarg** [comment]
I can imagine something like

```
foldJoin([i * [uint8] for i in range(1,50)])
```

**Comment by myitcv** [comment]
One way of doing this today is:

```
package x

x: [...int]
x: [1, 2]
// Not ok
// x: [1, 2, 3]
#checkXLength: len(x) < 3
#checkXLength: true
```

There have been discussions about how to make this more idiomatic (and indeed make the constraint on the length reusable) but nothing formalised as yet.

**Comment by mpvl** [comment, marked as answer]
There is a `list.MaxItems` in the `list` package. ~Builtin validators don't currently work correctly, however, as the evaluator doesn't handle non-monotonic constraints yet. This is addressed in the new evaluator work, which will handle such constraints.~

**Edit:** this is now supported (updated to reflect new support in new evaluator)

**Reply by myitcv** [reply]
> There is a `list.MaxItems` in the `list` package

I keep forgetting about that package....

**Reply by tnarg** [reply]
So you'd use list.MaxItems like this?
```
x: [...int]
#x_length_constraint: list.MaxItems(x, 3)
#x_length_constraint: true
```

**Reply by mpvl** [reply]
You would use it as a partial function like this:
```
a: list.MaxItems(3)
```
This is not officially supported yet, though, because CUE doesn't handle non-monotonic constraints properly. This is fixed in the new evaluator that I'm currently finishing up.

The way you suggest using it could work, I reckon.

**Reply by myitcv** [reply]
Update for anyone who sees/reads this discussion: this is now officially supported and works :)

**Comment by cueckoo** [comment]
This discussion has been migrated to https://github.com/cue-lang/cue/issues/401.

For more details about CUE's migration to a new home, please see https://github.com/cue-lang/cue/issues/1078.
